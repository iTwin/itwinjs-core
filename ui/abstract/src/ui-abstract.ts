/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

export * from "./ui-abstract/UiAbstract";
export * from "./ui-abstract/UiAdmin";
export * from "./ui-abstract/UiItemsManager";
export * from "./ui-abstract/UiItemsArbiter";

export * from "./ui-abstract/backstage/BackstageItem";
export * from "./ui-abstract/backstage/BackstageItemsManager";

export * from "./ui-abstract/common/KeyboardKey";

export * from "./ui-abstract/dialogs/DialogItem";
export * from "./ui-abstract/dialogs/UiLayoutDataProvider";
export * from "./ui-abstract/dialogs/UiDataProvider";

export * from "./ui-abstract/items/AbstractItemProps";
export * from "./ui-abstract/items/AbstractMenuItemProps";
export * from "./ui-abstract/items/ProvidedItem";
export * from "./ui-abstract/items/AbstractToolbarProps";
export * from "./ui-abstract/items/BadgeType";

export * from "./ui-abstract/items/ConditionalBooleanValue";
export * from "./ui-abstract/items/ConditionalStringValue";
export * from "./ui-abstract/items/RelativePosition";
export * from "./ui-abstract/items/ProvidedItem";
export * from "./ui-abstract/items/RelativePosition";
export * from "./ui-abstract/items/StageUsage";

export * from "./ui-abstract/properties/Description";
export * from "./ui-abstract/properties/EditorParams";
export * from "./ui-abstract/properties/PrimitiveTypes";
export * from "./ui-abstract/properties/Record";
export * from "./ui-abstract/properties/StandardEditorNames";
export * from "./ui-abstract/properties/StandardTypeNames";
export * from "./ui-abstract/properties/Value";

export * from "./ui-abstract/quantity/BaseQuantityDescription";
export * from "./ui-abstract/statusbar/StatusBarItem";
export * from "./ui-abstract/statusbar/StatusBarItemsManager";

export * from "./ui-abstract/toolbars/ToolbarItem";
export * from "./ui-abstract/toolbars/ToolbarItemsManager";

export * from "./ui-abstract/statusbar/StatusBarItem";
export * from "./ui-abstract/statusbar/StatusBarItemsManager";

export * from "./ui-abstract/utils/getClassName";
export * from "./ui-abstract/utils/IconSpecUtilities";
export * from "./ui-abstract/utils/callbacks";
export * from "./ui-abstract/utils/UiError";
export * from "./ui-abstract/utils/filter/charCode";
export * from "./ui-abstract/utils/filter/filters";
export * from "./ui-abstract/utils/filter/strings";

export * from "./ui-abstract/widget/AbstractWidgetProps";
export * from "./ui-abstract/widget/StagePanel";
export * from "./ui-abstract/widget/WidgetState";

/** @docs-package-description
 * The ui-abstract package contains abstractions for UI controls, such as toolbars, buttons and menus.
 * For more information, see [learning about ui-abstract]($docs/learning/ui/abstract/index.md).
 */
/**
 * @docs-group-description Backstage
 * Abstractions used by ui-framework package to create and manage the display Backstage menu items.
 */
/**
 * @docs-group-description Common
 * Common enums and functions used throughout the UI packages.
 */
/**
 * @docs-group-description Dialog
 * Interfaces and classes for generating UI items for Dialogs.
 */
/**
 * @docs-group-description Item
 * Classes for working with an Item in a Toolbar, Widget, Backstage or Context Menu
 */
/**
 * @docs-group-description Properties
 * Properties system for data input and formatting.
 */
/**
 * @docs-group-description StatusBar
 * Classes for creating and managing items in the status bar.
 */
/**
 * @docs-group-description Toolbar
 * Classes for creating and managing items in a toolbar.
 */
/**
 * @docs-group-description UiAdmin
 * Abstractions for UI controls, such as toolbars, buttons and menus and are callable from IModelApp.uiAdmin in imodeljs-frontend.
 */
/**
 * @docs-group-description UiItemsProvider
 * Interface for specifying UI items to be inserted at runtime.
 */
/**
 * @docs-group-description Utilities
 * Various utility classes for working with a UI.
 */
/**
 * @docs-group-description Widget
 * Classes for creating and providing Widgets.
 */
