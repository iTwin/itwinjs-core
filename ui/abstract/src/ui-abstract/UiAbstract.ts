/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Utilities
 */

import { Logger } from "@bentley/bentleyjs-core";
import { I18N } from "@bentley/imodeljs-i18n";
import { MessagePresenter } from "./notification/MessagePresenter";
import { getClassName } from "./utils/getClassName";
import { UiError } from "./utils/UiError";

/**
 * Entry point for static initialization required by various components used in the package.
 * @public
 */
export class UiAbstract {
  private static _initialized = false;
  private static _i18n?: I18N;
  private static _messagePresenter?: MessagePresenter;

  /**
   * Registers the I18N service namespace for UiAbstract
   * @param i18n The internationalization service created by the application.
   */
  public static async initialize(i18n: I18N): Promise<void> {
    if (UiAbstract._initialized) {
      Logger.logInfo(UiAbstract.loggerCategory(UiAbstract), `UiAbstract.initialize already called`);
      return;
    }

    UiAbstract._i18n = i18n;
    await UiAbstract._i18n.registerNamespace(UiAbstract.i18nNamespace).readFinished;
    UiAbstract._initialized = true;
  }

  /** Unregisters the UiAbstract internationalization service namespace */
  public static terminate() {
    if (UiAbstract._i18n)
      UiAbstract._i18n.unregisterNamespace(UiAbstract.i18nNamespace);
    UiAbstract._i18n = undefined;
    UiAbstract._initialized = false;
  }

  /** Determines if UiAbstract has been initialized */
  public static get initialized(): boolean { return UiAbstract._initialized; }

  /** The internationalization service created by the application. */
  public static get i18n(): I18N {
    if (!UiAbstract._i18n)
      throw new UiError(UiAbstract.loggerCategory(this), "UiAbstract not initialized");
    return UiAbstract._i18n;
  }

  /** The internationalization service namespace. */
  public static get i18nNamespace(): string {
    return "UiAbstract";
  }

  /** Calls i18n.translateWithNamespace with the "UiAbstract" namespace. Do NOT include the namespace in the key.
   * @internal
   */
  public static translate(key: string | string[]): string {
    return UiAbstract.i18n.translateWithNamespace(UiAbstract.i18nNamespace, key);
  }

  /** @internal */
  public static get packageName(): string {
    return "ui-abstract";
  }

  /** @internal */
  public static loggerCategory(obj: any): string {
    const className = getClassName(obj);
    const category = UiAbstract.packageName + (className ? `.${className}` : "");
    return category;
  }

  /** The MessagePresenter used to display messages. */
  public static get messagePresenter(): MessagePresenter {
    if (!UiAbstract._messagePresenter)
      throw new UiError(UiAbstract.loggerCategory(this), "UiAbstract.MessagePresenter not set");
    return UiAbstract._messagePresenter;
  }
  public static set messagePresenter(mp: MessagePresenter) {
    UiAbstract._messagePresenter = mp;
  }

}
