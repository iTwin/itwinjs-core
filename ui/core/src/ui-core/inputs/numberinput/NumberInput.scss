/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "../../base/base";

.core-number-input-container {
  position: relative;
  padding:  0;
  margin:   0;
  box-sizing: border-box;

  input {
    margin:       0;
    padding-right: 1.5em;
  }

  &.core-number-buttons-for-touch {
    input {
      padding-right: 3.5em;
    }
  }

  .core-number-input-buttons-container {
    position: absolute;
    top:      0;
    right:    0;
    width:    1.5em;
    height:   100%;
    @include uicore-centered;
    flex-direction: column;

    .core-number-input-button {
      font-size: .6em;
      outline: none;
      height: 100%;
      width: 100%;
      @include uicore-centered;
      &:hover {
        color: $buic-foreground-primary-tone;
        transition-duration: 150ms;
      }
      &.core-number-input-button-up {
        padding-top: 1px;
      }
      &.core-number-input-button-down {
        padding-top: 1px;
      }
    }

    &.core-number-buttons-for-touch {
      width:    3.5em;
      flex-direction: row-reverse;

      .core-number-input-button {
        font-size: 1em;
        &:hover {
          color: unset;
          transition-duration: unset;
        }

        &.core-number-input-button-up {
          padding-top: 2px;
        }
        &.core-number-input-button-down {
          padding-top: 2px;
        }
      }
    }
  }
}
