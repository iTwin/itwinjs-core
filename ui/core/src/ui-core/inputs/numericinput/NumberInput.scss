/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "../../base/base";

.core-numeric-input-container {
  position: relative;
  padding:  0;
  margin:   0;
  box-sizing: border-box;

  input {
    margin:       0;
    padding-right: 1.5em;
  }

  &.core-numeric-buttons-for-touch {
    input {
      padding-right: 3.5em;
    }
  }

  .core-numeric-input-buttons-container {
    position: absolute;
    top:      0;
    right:    0;
    width:    1.5em;
    height:   100%;
    @include uicore-centered;
    flex-direction: column;

    .core-numeric-input-button {
      font-size: .6em;
      outline: none;
      height: 100%;
      width: 100%;
      @include uicore-centered;
      &:hover {
        color: $buic-foreground-primary-tone;
        transition-duration: 150ms;
      }
      &.core-numeric-input-button-up {
        padding-top: 1px;
      }
      &.core-numeric-input-button-down {
        padding-top: 1px;
      }
    }

    &.core-numeric-buttons-for-touch {
      width:    3.5em;
      flex-direction: row-reverse;

      .core-numeric-input-button {
        font-size: 1em;
        &:hover {
          color: unset;
          transition-duration: unset;
        }

        &.core-numeric-input-button-up {
          padding-top: 2px;
        }
        &.core-numeric-input-button-down {
          padding-top: 2px;
        }
      }
    }
  }
}
