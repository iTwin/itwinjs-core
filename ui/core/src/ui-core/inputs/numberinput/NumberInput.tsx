/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Inputs
 */

import "./NumberInput.scss";
import classnames from "classnames";
import * as React from "react";
import { Input, InputProps } from "../Input";
import { WebFontIcon } from "../../icons/WebFontIcon";
import { SpecialKey } from "@bentley/ui-abstract";

/** Step function prototype for [[NumberInput]] component
 * @beta
 */
export type StepFunctionProp = number | ((direction: string) => number | undefined);

/** Properties for the [[NumberInput]] component
 * @beta
 */
export interface NumberInputProps extends Omit<InputProps, "min" | "max" | "step" | "onChange"> {
  /** Numeric value, set to `undefined` to show placeholder text */
  value?: number;
  /** CSS class name for the NumberInput component container div */
  containerClassName?: string;
  /** number or function	Number.MIN_SAFE_INTEGER */
  min?: number;
  /** number or function	defaults to Number.MAX_SAFE_INTEGER */
  max?: number;
  /** increment step value used while incrementing or decrementing (up/down buttons or arrow keys) defaults to 1. */
  step?: StepFunctionProp;
  /** number of decimal places, defaults to 0 */
  precision?: number;
  /** function parseFloat */
  parse?: ((value: string) => number | null | undefined);
  /** function optional formatting function that takes the number value and the internal formatted value in case function just adds prefix or suffix. */
  format?: (num: number | null | undefined, formattedValue: string) => string;
  /** Set to true to "snap" to the closest step value while incrementing or decrementing (up/down buttons or arrow keys). */
  snap?: boolean;
  /** Function to call when value is changed. */
  onChange?: (value: number | undefined, stringValue: string) => void;
  /** if true up/down buttons are shown larger and side by side */
  showTouchButtons?: boolean;
  /** Provides ability to return reference to HTMLInputElement */
  ref?: React.Ref<HTMLInputElement>;
}

const ForwardRefNumberInput = React.forwardRef<HTMLInputElement, NumberInputProps>(
  function ForwardRefNumberInput(props, ref) {
    const { containerClassName, value, min, max, precision, format, parse,
      onChange, onBlur, onKeyDown, step, snap, showTouchButtons, ...otherProps } = props;
    const currentValueRef = React.useRef(value);

    /**
   * Used internally to parse the argument x to it's numeric representation.
   * If the argument cannot be converted to finite number returns 0; If a
   * "precision" prop is specified uses it round the number with that
   * precision (no fixed precision here because the return value is float, not
   * string).
   */
    const parseInternal = React.useCallback((x: string) => {
      let n: number | undefined | null;

      if (parse)
        n = parse(x);
      if (undefined === n || null === n) {
        n = parseFloat(x);
        if (isNaN(n) || !isFinite(n)) {
          n = 0;
        }
      }

      const localPrecision = undefined === precision ? 10 : precision;
      const q = Math.pow(10, localPrecision);
      const localMin = undefined === min ? Number.MIN_SAFE_INTEGER : min;
      const localMax = undefined === max ? Number.MAX_SAFE_INTEGER : max;
      n = Math.min(Math.max(n, localMin), localMax);
      n = Math.round(n * q) / q;

      return n;
    }, [parse, precision, min, max]);

    /**
   * This is used internally to format a number to its display representation.
   * It will invoke the format function if one is provided.
   */
    const formatInternal = React.useCallback((num: number | undefined | null) => {
      const localPrecision = undefined === precision || null === precision ? 0 : precision;
      const str = undefined === num || null === num ? "" : num.toFixed(localPrecision);

      if (format)
        return format(num, str);

      return str;
    }, [format, precision]);

    const [formattedValue, setFormattedValue] = React.useState(() => formatInternal(value));

    // See if new initialValue props have changed since component mounted
    React.useEffect(() => {
      currentValueRef.current = value;
      const currentFormattedValue = formatInternal(currentValueRef.current);
      setFormattedValue(currentFormattedValue);
    }, [formatInternal, value]);

    const handleChange = React.useCallback((event: React.ChangeEvent<HTMLInputElement>) => {
      setFormattedValue(event.currentTarget.value);
    }, []);

    const updateValue = React.useCallback((newVal: number) => {
      const newFormattedVal = formatInternal(newVal);
      onChange && onChange(newVal, newFormattedVal);
      setFormattedValue(newFormattedVal);
    }, [onChange, formatInternal]);

    const updateValueFromString = React.useCallback((strValue: string) => {
      const newVal = parseInternal(strValue);
      updateValue(newVal);
    }, [parseInternal, updateValue]);

    const handleBlur = React.useCallback((event: React.FocusEvent<HTMLInputElement>) => {
      const newVal = parseInternal(event.target.value);
      onBlur && onBlur(event);
      updateValue(newVal);
    }, [parseInternal, updateValue, onBlur]);

    const getIncrementValue = React.useCallback((increment: boolean) => {
      if (typeof step === "function") {
        const stepVal = step(increment ? "up" : "down");
        return stepVal ? stepVal : 1;
      }

      return !step ? 1 : step;
    }, [step]);

    /**
   * The internal method that actually sets the new value on the input
   */
    const applyStep = React.useCallback((increment: boolean) => {
      const incrementValue = getIncrementValue(increment);

      let num = parseInternal(formattedValue) + (increment ? incrementValue : -incrementValue);
      if (snap) {
        num = Math.round(num / incrementValue) * incrementValue;
      }

      const localMin = undefined === min ? Number.MIN_SAFE_INTEGER : min;
      const localMax = undefined === max ? Number.MAX_SAFE_INTEGER : max;
      num = Math.min(Math.max(num, localMin), localMax);

      updateValue(num);
    }, [formattedValue, getIncrementValue, max, min, parseInternal, snap, updateValue]);

    const handleKeyDown = React.useCallback((event: React.KeyboardEvent<HTMLInputElement>) => {
    // istanbul ignore else
      if (event.key === SpecialKey.Enter) {
        updateValueFromString(event.currentTarget.value);
        event.preventDefault();
      } else if (event.key === SpecialKey.Escape) {
        setFormattedValue(formatInternal(currentValueRef.current));
        event.preventDefault();
      } else if (event.key === SpecialKey.ArrowDown) {
        applyStep(false);
        event.preventDefault();
      } else if (event.key === SpecialKey.ArrowUp) {
        applyStep(true);
        event.preventDefault();
      }
      onKeyDown && onKeyDown(event);
    }, [applyStep, formatInternal, updateValueFromString, onKeyDown]);

    const handleDownClick = React.useCallback((event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      applyStep(false);
      event.preventDefault();
    }, [applyStep]);

    const handleUpClick = React.useCallback((event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
      applyStep(true);
      event.preventDefault();
    }, [applyStep]);

    const containerClasses = classnames("core-number-input-container", containerClassName, showTouchButtons && "core-number-buttons-for-touch");
    return (
      <div className={containerClasses} >
        <Input ref={ref} value={formattedValue} onChange={handleChange} onKeyDown={handleKeyDown} onBlur={handleBlur} {...otherProps} />
        <div className={classnames("core-number-input-buttons-container", showTouchButtons && "core-number-buttons-for-touch")}>
          { /* eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */}
          <div className="core-number-input-button core-number-input-button-up" tabIndex={-1} onClick={handleUpClick}>
            <WebFontIcon iconName="icon-caret-up" />
          </div>
          { /* eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/click-events-have-key-events */}
          <div className="core-number-input-button core-number-input-button-down" tabIndex={-1} onClick={handleDownClick}>
            <WebFontIcon iconName="icon-caret-down" />
          </div>
        </div>
      </div>
    );
  }
);

/** Input component for numbers with up and down buttons to increment and decrement the value.
 * @beta
 */
export const NumberInput: (props: NumberInputProps) => JSX.Element | null = ForwardRefNumberInput;
