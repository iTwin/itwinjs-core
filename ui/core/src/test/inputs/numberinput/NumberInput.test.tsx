/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { SpecialKey } from "@bentley/ui-abstract";
import { fireEvent, render } from "@testing-library/react";
import { expect } from "chai";
import * as React from "react";
import * as sinon from "sinon";
import { NumberInput } from "../../../ui-core/inputs/numberinput/NumberInput";

// cSpell:ignore decrementor

function parseDollar(stringValue: string) {
  const noDollarSign = stringValue.replace(/^\$/, "");
  let n = parseFloat(noDollarSign);
  if (isNaN(n) || !isFinite(n))
    n = 0;
  return n;
}

function formatDollar(num: number | undefined | null, fallback: string) {
  if (undefined === num || null === num)
    return fallback;

  return `$${num.toFixed(2)}`;
}

function exoticStep(direction: string) {
  if (direction === "up")
    return .5;
  return .1;
}

function undefinedStepFunction(_direction: string) {
  return undefined;
}

describe("<NumberInput - React Testing Library />", () => {
  it("value should update with up/down buttons", () => {
    const spyMethod = sinon.spy();
    let updatedValue: number | undefined = 5;
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      updatedValue = v;
      spyMethod();
    };

    const wrapper = render(<NumberInput value={1} step={undefined} onChange={handleChange} />);
    const input = wrapper.container.querySelector("input") as HTMLInputElement;
    expect(input.value).to.eq("1");

    wrapper.rerender(<NumberInput value={5} step={undefined} onChange={handleChange} />);
    expect(input.value).to.eq("5");

    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    const decrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-down");
    expect(decrementor).not.to.be.null;

    fireEvent.click(incrementor!);
    expect(input.value).to.eq("6");
    spyMethod.calledOnce.should.true;
    expect(updatedValue).to.eq(6);

    fireEvent.click(decrementor!);
    expect(updatedValue).to.eq(5);
  });

  it("steps correctly with undefined step", () => {
    let value: number | undefined = 0;
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      value = v;
    };
    const wrapper = render(<NumberInput value={value} step={undefined} onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    expect(value).to.eq(1);
  });

  it("steps correctly with number step", () => {
    let value: number | undefined = 0;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput value={value} step={5} onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(5);
  });

  it("steps correctly with decimal step", () => {
    let value: number | undefined = 1.23;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput precision={2} value={value} step={.25} onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(1.48);
  });

  it("properly handle max", () => {
    let value: number | undefined = 0;
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      value = v;
    };
    const wrapper = render(<NumberInput value={value} step={1} max={5} onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);  // 1
    fireEvent.click(incrementor!);  // 2
    fireEvent.click(incrementor!);  // 3
    fireEvent.click(incrementor!);  // 4
    fireEvent.click(incrementor!);  // 5
    fireEvent.click(incrementor!);  // 6 => 5
    expect(value).to.eq(5);
  });

  it("properly handle MAX_SAFE_INTEGER value", () => {
    let value: number | undefined = Number.MAX_SAFE_INTEGER;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput value={value} step={1} onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(Number.MAX_SAFE_INTEGER);
  });

  it("properly handle min", () => {
    let value: number | undefined = 0;
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      value = v;
    };
    const wrapper = render(<NumberInput value={value} step={1} min={-5} onChange={handleChange} />);
    const decrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-down");
    expect(decrementor).not.to.be.null;
    fireEvent.click(decrementor!);  // -1
    fireEvent.click(decrementor!);  // -2
    fireEvent.click(decrementor!);  // -3
    fireEvent.click(decrementor!);  // -4
    fireEvent.click(decrementor!);  // -5
    fireEvent.click(decrementor!);  // -6 => -5
    expect(value).to.eq(-5);
  });

  it("properly handle MIN_SAFE_INTEGER value", () => {
    let value: number | undefined = Number.MIN_SAFE_INTEGER;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput value={value} step={1} onChange={handleChange} />);
    const decrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-down");
    expect(decrementor).not.to.be.null;
    fireEvent.click(decrementor!);
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(Number.MIN_SAFE_INTEGER);
  });

  it("steps correctly with decimal step and snap", () => {
    let value: number | undefined = 1.23;
    let formattedValue: string = "";
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, stringValue: string): void => {
      spyMethod();
      value = v;
      formattedValue = stringValue;
    };

    const wrapper = render(<NumberInput precision={2} value={value} step={.25} snap onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(1.5);
    expect(formattedValue).to.eq("1.50");
  });

  it("steps correctly when placeholder is used", () => {
    let value: number | undefined = 0;
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      value = v;
    };
    const wrapper = render(<NumberInput placeholder="Enter Text" step={1} onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    expect(value).to.eq(1);
    const decrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-down");
    expect(decrementor).not.to.be.null;
    fireEvent.click(decrementor!);
    expect(value).to.eq(0);
  });

  it("steps correctly with function step +.5/-.1", () => {
    let value: number | undefined = 0;
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      value = v;
    };
    // Note: requires precision to avoid round off during incrementing.
    const wrapper = render(<NumberInput value={value} precision={1} step={exoticStep} onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    expect(value).to.eq(.5);
    const decrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-down");
    expect(decrementor).not.to.be.null;
    fireEvent.click(decrementor!);
    expect(value).to.eq(.4);
  });

  it("steps correctly when step function return undefined", () => {
    let value: number | undefined = 0;
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      value = v;
    };
    // Note: requires precision to avoid round off during incrementing.
    const wrapper = render(<NumberInput value={value} precision={1} step={undefinedStepFunction} onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    expect(value).to.eq(1);
    const decrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-down");
    expect(decrementor).not.to.be.null;
    fireEvent.click(decrementor!);
    expect(value).to.eq(0);
  });

  it("should increment/decrement value on Up/Down Arrow", () => {
    let value: number | undefined = 1.23;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput precision={2} value={value} step={.25} onChange={handleChange} />);
    const input = wrapper.container.querySelector("input");
    expect(input).not.to.be.null;
    fireEvent.keyDown(input!, { key: SpecialKey.ArrowUp });
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(1.48);

    spyMethod.resetHistory();
    fireEvent.keyDown(input!, { key: SpecialKey.ArrowDown });
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(1.23);
  });

  it("should update value on enter", () => {
    let value: number | undefined = 1.23;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput precision={2} value={value} step={.25} onChange={handleChange} />);
    const input = wrapper.container.querySelector("input");
    expect(input).not.to.be.null;
    fireEvent.change(input!, { target: { value: "22.3" } });
    fireEvent.keyDown(input!, { key: SpecialKey.Enter });
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(22.3);
  });

  it("should update value on blur", () => {
    let value: number | undefined = 1.23;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput precision={2} value={value} step={.25} onChange={handleChange} />);
    const input = wrapper.container.querySelector("input");
    expect(input).not.to.be.null;
    input?.focus();
    fireEvent.change(input!, { target: { value: "22.3" } });
    input?.blur();
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(22.3);
  });

  it("should reset value on ESC", () => {
    let value: number | undefined = 1.23;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput precision={2} value={value} step={.25} onChange={handleChange} />);
    const input = wrapper.container.querySelector("input");
    expect(input).not.to.be.null;
    const originalValue = (input as HTMLInputElement).value;
    fireEvent.change(input!, { target: { value: "22.3" } });
    expect((input as HTMLInputElement).value).to.eq("22.3");
    fireEvent.keyDown(input!, { key: SpecialKey.Escape });
    spyMethod.notCalled.should.be.true;
    expect((input as HTMLInputElement).value).to.eq(originalValue);
  });

  it("should reset value to 0 when invalid text is entered", () => {
    let value: number | undefined = 1.23;
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, _stringValue: string): void => {
      spyMethod();
      value = v;
    };
    const wrapper = render(<NumberInput precision={2} value={value} step={.25} onChange={handleChange} />);
    const input = wrapper.container.querySelector("input");
    expect(input).not.to.be.null;
    fireEvent.change(input!, { target: { value: "abc" } });
    expect((input as HTMLInputElement).value).to.eq("abc");
    fireEvent.keyDown(input!, { key: SpecialKey.Enter });
    spyMethod.calledOnce.should.be.true;
    expect(value).to.eq(0);
  });

  it("renders for touch correctly", () => {
    const wrapper = render(<NumberInput value={0} showTouchButtons />);
    const mainContainer = wrapper.container.querySelector("div.core-number-input-container.core-number-buttons-for-touch");
    expect(mainContainer).not.to.be.null;
    const buttonContainer = wrapper.container.querySelector("div.core-number-input-buttons-container.core-number-buttons-for-touch");
    expect(buttonContainer).not.to.be.null;
  });

  it("processes parse and format functions correctly", () => {
    let value: number | undefined = 1.23;
    let formattedValue: string = "";
    const spyMethod = sinon.spy();
    const handleChange = (v: number | undefined, stringValue: string): void => {
      spyMethod();
      value = v;
      formattedValue = stringValue;
    };

    const wrapper = render(<NumberInput format={formatDollar} parse={parseDollar} precision={2} value={value} step={.25} snap onChange={handleChange} />);
    const incrementor = wrapper.container.querySelector("div.core-number-input-button.core-number-input-button-up");
    expect(incrementor).not.to.be.null;
    fireEvent.click(incrementor!);
    spyMethod.calledOnce.should.true;
    expect(value).to.eq(1.5);
    expect(formattedValue).to.eq("$1.50");

    const input = wrapper.container.querySelector("input");
    expect(input).not.to.be.null;
    fireEvent.change(input!, { target: { value: "15.3" } });
    expect((input as HTMLInputElement).value).to.eq("15.3");
    fireEvent.keyDown(input!, { key: SpecialKey.Enter });
    expect(formattedValue).to.eq("$15.30");

    fireEvent.change(input!, { target: { value: "$2.25" } });
    expect((input as HTMLInputElement).value).to.eq("$2.25");
    fireEvent.keyDown(input!, { key: SpecialKey.Enter });
    expect(formattedValue).to.eq("$2.25");
  });

});

