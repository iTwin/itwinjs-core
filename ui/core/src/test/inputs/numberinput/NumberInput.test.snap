// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput should render correctly 1`] = `
<div
  className="core-number-input-container"
>
  <ForwardRef(Input)
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    value=""
  />
  <div
    className="core-number-input-buttons-container"
  >
    <div
      className="core-number-input-button core-number-input-button-up"
      onClick={[Function]}
      tabIndex={-1}
    >
      <WebFontIcon
        iconName="icon-caret-up"
      />
    </div>
    <div
      className="core-number-input-button core-number-input-button-down"
      onClick={[Function]}
      tabIndex={-1}
    >
      <WebFontIcon
        iconName="icon-caret-down"
      />
    </div>
  </div>
</div>
`;
